### PR TITLE
KOGITO-1046 - Data Index performance tests

### PR DESCRIPTION
**Pull request by MarianMacik**
Many thanks for submitting your Pull Request :heart: :racing_car:!
[KOGITO-1046](https://issues.redhat.com/browse/KOGITO-1046)

Feature file is not ready yet (will be finished in another PR) as I still have to tune up Kafka and Kogito service to get to the desired numbers. However, I chose to do this as a part of another JIRA ([KOGITO-3742](https://issues.redhat.com/browse/KOGITO-3742)), which will be purely focused on tuning the parameters and infrastructure. Therefore you can ignore the enclosed feature file, it has the tag `@performance`, so it won't run by default.

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

File: test/steps/image_registry.go [hunk 1]
```diff
@@ -24,26 +24,46 @@ import (
...
 
+/*
+	DataTable for Maven:
+	| -Dproperty=value |
+*/
+
 // registerImageRegistrySteps register all existing image registry steps
 func registerImageRegistrySteps(ctx *godog.ScenarioContext, data *Data) {
 	ctx.Step(`^Local example service "([^"]*)" is built by Maven using profile "([^"]*)" and deployed to runtime registry$`, data.localServiceBuiltByMavenWithProfileAndDeployedToRuntimeRegistry)
+	ctx.Step(`^Local example service "([^"]*)" is built by Maven using profile "([^"]*)" and options and deployed to runtime registry:$`, data.localServiceBuiltByMavenWithProfileAndOptionsAndDeployedToRuntimeRegistry)
```
Comment (radtriste) [inline test/steps/image_registry.go:38]:
```suggestion
	ctx.Step(`^Local example service "([^"]*)" is built by Maven using profile "([^"]*)" and deployed to runtime registry with maven options:$`, data.localServiceBuiltByMavenWithProfileAndOptionsAndDeployedToRuntimeRegistry)
```
I think we should let the table information be the last part of the step, so we know easily what it is for

File: test/steps/image_registry.go [hunk 2]
```diff
@@ -114,13 +134,13 @@ func getKogitoImageTag(projectLocation, mavenProfiles string) (string, error) {
...
-	kogitoImageName := getKogitoImageName(projectLocation, mavenProfiles)
+	kogitoImageName := getKogitoImageName(projectLocation, mavenProfiles, options)
 	buildImageTag := fmt.Sprintf("%s/%s/%s:%s", runtimeApplicationImageRegistry, runtimeApplicationImageNamespace, kogitoImageName, runtimeApplicationImageVersion)
 	return buildImageTag, nil
 }
 
-// Returns Kogito image name in the form of "<project location base>(-<maven profile>)*"
-func getKogitoImageName(projectLocation, mavenProfiles string) string {
+// Returns Kogito image name in the form of "<project location base>(-<maven profile>)*(-<option>)*"
+func getKogitoImageName(projectLocation, mavenProfiles string, options []string) string {
```
Comment (radtriste) [inline test/steps/image_registry.go:143]: Please call them mavenOptions ?

File: test/framework/graphql.go [hunk 3]
```diff
@@ -47,6 +82,11 @@ func WaitForSuccessfulGraphQLRequestUsingAccessToken(namespace, uri, path, query
 
 // ExecuteGraphQLRequest executes a GraphQL query
 func ExecuteGraphQLRequest(namespace, uri, path, query, bearerToken string, response interface{}) error {
+	return ExecuteGraphQLRequestWithLogging(namespace, uri, path, query, bearerToken, response, true)
+}
+
+// ExecuteGraphQLRequestWithLogging executes a GraphQL query with possibility of logging each request
+func ExecuteGraphQLRequestWithLogging(namespace, uri, path, query, bearerToken string, response interface{}, logResponse bool) error {
```
[the remark this answers is not in the document]
Comment (MarianMacik) [reply test/framework/graphql.go:89]: I was trying to make it as flexible as possible so we don't have to call it with `true` each time. As this option to turn the logging off is pretty much just for performance tests when we are executing a lot of requests and I didn't want it to be populating logs with thousands of responses. Do we still want to provide logging true or false?
I can do `ExecuteGraphQLRequestWithLogging` which will call true for logging and then `ExecuteGraphQLRequestWithLoggingOption` which will be the actual implementation.

File: test/framework/graphql.go [hunk 4]
```diff
@@ -28,6 +30,39 @@ func WaitForSuccessfulGraphQLRequest(namespace, uri, path, query string, timeout
...
+		func() (bool, error) {
+			for queried := 0; queried < totalSize; {
+				query := strings.ReplaceAll(query, "$offset", strconv.Itoa(queried))
+				GetLogger(namespace).Infof("Querying page no. %d of %d using query: %s", queried/pageSize + 1, totalSize/pageSize, query)
+				err := ExecuteGraphQLRequestWithLogging(namespace, uri, path, query, graphQLNoAuthToken, response, false)
+				if err != nil {
+					return false, err
+				}
+				if analyzeResponse != nil {
+					if success, err := analyzeResponse(response); !success || err != nil {
```
[the remark this answers is not in the document]
Comment (MarianMacik) [reply test/framework/graphql.go:51]: Hmm, what is confusing with the current naming? Maybe I was thinking about `afterEachResponse` and `afterFullRequest`? Is it better?

File: test/steps/image_registry.go (hunk 2 above)
Comment (MarianMacik) [reply test/steps/image_registry.go:143]: OK, but inside MavenCommand they can be still called Options, right? That would probably make sense. Wdyt?

File: test/steps/image_registry.go (hunk 1 above)
Comment (MarianMacik) [reply test/steps/image_registry.go:38]: +1

File: test/framework/graphql.go [hunk 5]
```diff
@@ -28,6 +30,39 @@ func WaitForSuccessfulGraphQLRequest(namespace, uri, path, query string, timeout
 	return WaitForSuccessfulGraphQLRequestUsingAccessToken(namespace, uri, path, query, graphQLNoAuthToken, timeoutInMin, response, analyzeResponse)
 }
 
+// WaitForSuccessfulGraphQLRequestUsingPagination waits for a GraphQL request with pagination to be successful.
```
[the remark this answers is not in the document]
Comment (MarianMacik) [reply test/framework/graphql.go:33]: +1

File: test/framework/graphql.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (radtriste) [reply test/framework/graphql.go:51]: yes, even better :)

File: test/framework/graphql.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (radtriste) [reply test/framework/graphql.go:89]: yes, seems good

File: test/steps/image_registry.go (hunk 2 above)
Comment (radtriste) [reply test/steps/image_registry.go:143]: yes, for MavenCommand that can stay as `Options`
for the rest, options is not explicit enough to be Maven related, that was my suggestion :)

File: test/steps/image_registry.go [hunk 6]
```diff
@@ -24,26 +24,46 @@ import (
...
 
 // Build local service and deploy it to registry if the registry doesn't contain such image already
 func (data *Data) localServiceBuiltByMavenWithProfileAndDeployedToRuntimeRegistry(contextDir, profile string) error {
+	return data.localServiceBuiltByMavenWithProfileAndOptionsAndDeployedToRuntimeRegistry(contextDir, profile, nil)
+}
+
+// Build local service and deploy it to registry if the registry doesn't contain such image already
+func (data *Data) localServiceBuiltByMavenWithProfileAndOptionsAndDeployedToRuntimeRegistry(contextDir, profile string, table *godog.Table) error {
+	var options []string
+	if len(table.Rows) > 0 {
```
Comment (sutaakar) [inline test/steps/image_registry.go:49]:
```suggestion
	if table!= nil &&  len(table.Rows) > 0 {
```

Comment (MarianMacik) [reply test/steps/image_registry.go:49]: good catch! :)

File: test/steps/image_registry.go (hunk 1 above)
Comment (MarianMacik) [reply test/steps/image_registry.go:38]: I originally didn't want it that way as it suggests that it is deployed to registry with these options whereas it is just built with these options, but probably still better than the current naming :)

File: test/steps/image_registry.go [hunk 7]
```diff
@@ -24,26 +24,46 @@ import (
...
 
+/*
+	DataTable for Maven:
+	| -Dproperty=value |
+*/
+
 // registerImageRegistrySteps register all existing image registry steps
 func registerImageRegistrySteps(ctx *godog.ScenarioContext, data *Data) {
 	ctx.Step(`^Local example service "([^"]*)" is built by Maven using profile "([^"]*)" and deployed to runtime registry$`, data.localServiceBuiltByMavenWithProfileAndDeployedToRuntimeRegistry)
+	ctx.Step(`^Local example service "([^"]*)" is built by Maven using profile "([^"]*)" and deployed to runtime registry with options:$`, data.localServiceBuiltByMavenWithProfileAndDeployedToRuntimeRegistryWithOptions)
```
Comment (radtriste) [inline test/steps/image_registry.go:38]: I was suggesting `with Maven options`, so you know what are the options for ;)

File: test/steps/image_registry.go [hunk 8]
```diff
@@ -24,26 +24,46 @@ import (
...
 }
 
 // Build local service and deploy it to registry if the registry doesn't contain such image already
 func (data *Data) localServiceBuiltByMavenWithProfileAndDeployedToRuntimeRegistry(contextDir, profile string) error {
+	return data.localServiceBuiltByMavenWithProfileAndDeployedToRuntimeRegistryWithOptions(contextDir, profile, nil)
+}
+
+// Build local service and deploy it to registry if the registry doesn't contain such image already
+func (data *Data) localServiceBuiltByMavenWithProfileAndDeployedToRuntimeRegistryWithOptions(contextDir, profile string, table *godog.Table) error {
+	var options []string
```
Comment (radtriste) [inline test/steps/image_registry.go:48]: here we should call it `mavenOptions` IMO

Comment (MarianMacik) [reply test/steps/image_registry.go:48]: Well, we don't call them mavenProfiles here it as well, so was keeping it consistent :) Is that OK? We have profiles and options and mavenProfiles and mavenOptions.

Comment (radtriste) [reply test/steps/image_registry.go:48]: oki good

**Comment by MarianMacik** [comment]
Updated :)

**Review by radtriste — APPROVED** [review]

**Review by sutaakar — APPROVED** [review]

**Pull request merged**